File: src/initial_translation/extrapolation.py
# Python modules
import jax.numpy as jnp
from jax import grad, debug
import matplotlib.pyplot as plt

# Application modules
from SPRE_stepwise import SPRE_stepwise
from GRE_stepwise import GRE_stepwise

def extrapolation(X, Y, options=None):
    """
    Extrapolation to estimate f(0) from input-output training data (X, Y).

    Parameters:
        X : np.ndarray of shape (n_train, d)
            Training input vectors.
        Y : np.ndarray of shape (n_train,)
            Training scalar outputs.
        options : dict, optional
            Extrapolation options with keys:
                - "name"   : str, one of {"MRE", "GRE", "SPRE"} (default: "SPRE")
                - "k_name" : str, one of {"Gaussian", "GaussianARD", "Matern1/2", "Matern3/2", "white"} (default: "white")
                - "plot"   : bool, whether to plot LOOCV results (default: True)

    Returns:
        out : dict
            A dictionary containing:
                - "mu"     : predictive mean for f(0)
                - "var"    : predictive variance (if available)
                - "mu_cv"  : LOOCV means (optional)
                - "var_cv" : LOOCV variances (optional)
    """

    # Default options
    if options is None:
        options = {}
    name = options.get("name", "SPRE")
    k_name = options.get("k_name", "white")
    plot = options.get("plot", True)

    # Select extrapolation method
    if name == "MRE":
        raise NotImplementedError("MRE extrapolation is not implemented yet.")
    elif name == "GRE":
        out = GRE_stepwise(X, Y, k_name)
    elif name == "SPRE":
        out = SPRE_stepwise(X, Y, k_name)
    else:
        raise ValueError(f"Unknown extrapolation method: {name}")

    errors = jnp.sqrt(out["var_cv"])

    # Plot LOOCV fit if applicable
    if plot and name != "MRE" and "mu_cv" in out and "var_cv" in out:
        n_train = X.shape[0]
        plt.figure()
        plt.errorbar(jnp.arange(1, n_train + 1), out["mu_cv"].flatten(), yerr = errors, fmt='bo', label='predicted')
        plt.scatter(jnp.arange(1, n_train + 1), Y, color='k', marker='x', label='actual')
        plt.xticks(jnp.arange(1, n_train + 1))
        plt.xlabel(r'$i$')
        plt.ylabel(r'$f(\mathbf{x}_i)$')
        plt.title(f"Leave-one-out cross validation ({name})")
        plt.legend()
        plt.show()

    return out

File: matlab_code/extrapolation.m
%% Extrapolation
% This code extrapolates the value f(0) of a function f : R^d -> R based 
% on training data y_i = f(x_i).
%
% Inputs:
% X       = n_train x d, training (vector) inputs
% Y       = n_train x 1, training (scalar) outputs
% options = options for extrapolation method 
%
% Options:
% options.name      = string, "MRE" (multivariate Richardson extrapolation)
%                             "GRE" (Gauss-Richardson extrapolation)
%                          or "SPRE" (default; sparse probabilistic Richardson extrapolation)
% options.k_name    = string, "Gaussian"
%                             "GaussianARD"
%                             "Matern1/2"
%                             "Matern3/2" 
%                          or "white" (default)
% options.max_order = integer, maximum polynomial order
% options.plot      = logical, plot LOOCV fit if probabilistic method used (default)
%
% Outputs:
% out.mu  = predictive mean for f(0)
% out.var = predictive variance for f(0), if probabilistic method used
% (depending on the extrapolation method, other outputs may be present)
%
% Examples:
% d = 2; % data dimension
% n_train = 5; % number of training data
% X = rand(n_train,d); % training inputs
% Y = 3 + sin(X(:,1)) + sin(X(:,2)); % training outputs
% 
% out = extrapolation(X,Y,[]); % extrapolate with default options
% disp("predict f(0) = " + out.mu + "+/-" + sqrt(out.var)) 
%
% options.name = "SPRE"; % use SPRE
% options.k_name = "Gaussian"; % use Gaussian kernel
% options.plot = true; % plot LOOCV fit
% out = extrapolation(X,Y,options); % extrapolate with custom options
% disp("predict f(0) = " + out.mu + "+/-" + sqrt(out.var)) 

function out = extrapolation(X,Y,options)

% default options, if not provided
if ~isfield(options,"name")
    options.name = "SPRE";
end
if ~isfield(options,"k_name")
    options.k_name = "white";
end
if ~isfield(options,"plot")
    options.plot = true;
end

% call the requested extrapolation method
if strcmp(options.name,"MRE")
    error("not coded yet")
elseif strcmp(options.name,"GRE")
    out = GRE_stepwise(X,Y,options.k_name,options.max_order);  %%%%%%%%% NEW
elseif strcmp(options.name,"SPRE")
    out = SPRE_stepwise(X,Y,options.k_name,options.max_order);  %%%%%%%%% NEW
end

% plot LOOCV fit
n_train = size(X,1);
if options.plot && ~strcmp(options.name,"MRE")
    figure()
    errorbar(1:n_train,out.mu_cv,out.var_cv.^(1/2),'bo'); hold on
    scatter(1:n_train,Y,'kx')
    xticks(1:n_train)
    xlabel('$i$','Interpreter','latex')
    ylabel('$f(\mathbf{x}_i)$','Interpreter','latex')
    legend({'predicted','actual'})
    title("Leave-one-out cross validation (" + options.name + ")")
end

end



